### ISSUE TITLE
PATH override in repository_ctx.execute leads to key error

### ISSUE THREAD

**Issue by abergmeier** [bazelbuild/bazel#1472]
When I have the following:

    env = dict()
    env["PATH"] = "foo"

    repository_ctx.execute([
        "sh",
      ],
      600,
      env
    )

I get an error:

    Multiple entries with the same key: PATH=foo and PATH=systemstuff

Is overriding possible at all?

**Comment by aehlig** [comment]
> I get an error:
> 
>     Multiple entries with the same key: PATH=foo and PATH=systemstuff
> 
> Is overriding possible at all?

Just a side remark: For environment variables, especially the PATH one,
we're eventually moving to a new design, see
http://www.bazel.io/docs/designs/2016/06/21/environment.html so the
current state of affairs is a bit in flux anyway.

-- 
Klaus Aehlig
Google Germany GmbH, Erika-Mann-Str. 33, 80636 Muenchen
Registergericht und -nummer: Hamburg, HRB 86891
Sitz der Gesellschaft: Hamburg
Geschaeftsfuehrer: Matthew Scott Sucherman, Paul Terence Manicle

**Comment by abergmeier** [comment]
Sounds interesting. Any ETA?
As a side side remark: We need `PATH` not only in the rule_ctx but also in repository_ctx.
And maybe I missed it but will externals use the root (WORKSPACE) rc, too?

**Comment by aehlig** [comment]
> Sounds interesting. Any ETA?

Our current plan is to implement it by end of September, but don't hold your breath; it might well slip a month or two.

> And maybe I missed it but will externals use the root (WORKSPACE) rc, too?

The design as it currently is will use the current rc-file mechanism. That is, rc-files are read
at bazel invocation (so, before any workspace file is read) and apply to any action that this bazel invocation carries out.

**Comment by aehlig** [comment]
Just to clarify: would a well-document current, or future status as per the design work for you? If not, could you give an example of a use case that would not work?

CC @philwo

**Comment by abergmeier** [comment]
So with regards to the initial issue - I think the problem lies directly in `ImmutableMap.Builder`. Working around that "limitation" would be relatively easy in `SkylarkRepositoryContext`.
With @damienmg bragging ;) that he will make `SkylarkRepositoryContext` shinier and better soonish I wonder whether I should fix this.
(Currently) I could sit this one out for a month if necessary.

**Comment by abergmeier** [comment]
@aehlig With our current requirement - yes this should work (we only need paths to inside Bazel).
That being said it seems to me that stuff like appending to variables is not possible with this approach, right?
So the infamous `LD_LIBRARY_PATH=myjunk:$LD_LIBRARY_PATH` can not be mapped with rcs?

**Comment by aehlig** [comment]
> With our current requirement - yes this should work

Good to hear. So we proceed with the current plan.

> So the infamous `LD_LIBRARY_PATH=myjunk:$LD_LIBRARY_PATH` can not be mapped with rcs?

Yes, not directly. But no one prevents you from doing a
```
echo common --action-env=LD_LIBRARY_PATH=myjunk:$LD_LIBRARY_PATH >> .bazelrc
```
next to the `WORKSPACE` immediately after checking out the repository in question.

**Comment by aehlig** [comment]
Assigning a low priority to the documentation issue, as it probably makes more sense to spend effort into implementing the new design than documenting the old status quo.

**Comment by damienmg** [comment]
This has nothing to do with sandboxing and this was fixed at head IIRC.

**Comment by kchodorow** [comment]
Which bug?

**Comment by abergmeier** [comment]
Interesting, my example does not even work with 0.3.0 yet.
Anyway, with latest master it works. Do not know who fixed it but thanks :+1:

**Comment by damienmg** [comment]
25a98e16731a2a0cbb59449a6e7d395bd204d04c
I was the one fixing it actually :)
